File: csharp/Betfair.ESAClient/Betfair.ESAClient/Auth/AppKeyAndSessionProvider.cs
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Net;
using System.Runtime.Serialization;
using System.Security.Authentication;
using System.Text;
using System.Threading.Tasks;

namespace Betfair.ESAClient.Auth {
    /// <summary>
    /// Utility class to provide a session & token via identity SSO
    /// </summary>
    public class AppKeyAndSessionProvider {
        private string _appkey;
        private string _host;
        private string _password;
        private string _username;

        private AppKeyAndSession _session;

        public const string SSO_HOST_COM = "identitysso.betfair.com";
        public const string SSO_HOST_IT = "identitysso.betfair.it";
        public const string SSO_HOST_ES = "identitysso.betfair.es";

        public AppKeyAndSessionProvider(string ssoHost, string appkey, string username, string password) {
            _host = ssoHost;
            _appkey = appkey;
            _username = username;
            _password = password;
            Timeout = TimeSpan.FromSeconds(30);
            //4hrs is normal expire time
            SessionExpireTime = TimeSpan.FromHours(3);
        }

        /// <summary>
        /// AppKey being used
        /// </summary>
        public string Appkey {
            get { return _appkey; }
        }

        /// <summary>
        /// Session expire time (default 3hrs)
        /// </summary>
        public TimeSpan SessionExpireTime { get; set; }

        /// <summary>
        /// Specifies the timeout
        /// </summary>
        public TimeSpan Timeout { get; set; }


        /// <summary>
        /// Constructs a new session token via identity SSO.
        /// Note: These are not cached.
        /// </summary>
        /// <exception cref="InvalidCredentialException">Thrown if authentication response is fail</exception>
        /// <exception cref="IOException">Thrown if authentication call fails</exception>
        /// <returns></returns>
        public AppKeyAndSession GetOrCreateNewSession() {
            if (_session != null) {
                //have a cached session - is it expired
                if ((_session.CreateTime + SessionExpireTime) > DateTime.UtcNow) {
                    Trace.TraceInformation("SSO Login - session not expired - re-using");
                    return _session;
                }
                else {
                    Trace.TraceInformation("SSO Login - session expired");
                }
            }

            Trace.TraceInformation("SSO Login host={0}, appkey={1}, username={2}",
                _host,
                _appkey,
                _username);
            SessionDetails sessionDetails;
            try {
                string uri = string.Format("https://{0}/api/login?username={1}&password={2}",
                    _host,
                    _username,
                    _password);

                HttpWebRequest loginRequest = (HttpWebRequest) WebRequest.Create(uri);
                loginRequest.Headers.Add("X-Application", _appkey);
                loginRequest.Accept = "application/json";
                loginRequest.Method = "POST";
                loginRequest.Timeout = (int) Timeout.TotalMilliseconds;
                WebResponse thePage = loginRequest.GetResponse();
                using (StreamReader reader = new StreamReader(thePage.GetResponseStream())) {
                    string response = reader.ReadToEnd();
                    Trace.TraceInformation("{0}: Response: {1}", _host, response);
                    sessionDetails = JsonConvert.DeserializeObject<SessionDetails>(response);
                }
            }
            catch (Exception e) {
                throw new IOException("SSO Authentication - call failed:", e);
            }

            //got a response - decode
            if (sessionDetails != null && "SUCCESS".Equals(sessionDetails.status)) {
                _session = new AppKeyAndSession(_appkey, sessionDetails.token);
            }
            else {
                throw new InvalidCredentialException("SSO Authentication - response is fail: " + sessionDetails.error);
            }

            return _session;
        }

        /// <summary>
        /// Expires cached token
        /// </summary>
        public void ExpireTokenNow() {
            Trace.TraceInformation("SSO Login - expiring session token now");
            _session = null;
        }
    }

    class SessionDetails {
        public string token;
        public string product;
        public string status;
        public string error;
    }
}

File: java/client/src/main/java/com/betfair/esa/client/auth/AppKeyAndSessionProvider.java
package com.betfair.esa.client.auth;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.sun.jersey.api.client.Client;
import com.sun.jersey.api.client.ClientResponse;
import com.sun.jersey.api.client.WebResource;
import java.io.IOException;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.time.Clock;
import java.time.Duration;
import java.time.Instant;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/** Created by hoszua on 07/07/2016. */
public class AppKeyAndSessionProvider {

    private final Logger logger = LoggerFactory.getLogger(getClass().getName());

    private final Duration timeout;
    private final Duration sessionExpireTime;

    private final String appkey;
    private final String host;
    private final String password;
    private final String username;

    private AppKeyAndSession session;

    public static final String SSO_HOST_COM = "identitysso.betfair.com";
    public static final String SSO_HOST_IT = "identitysso.betfair.it";
    public static final String SSO_HOST_ES = "identitysso.betfair.es";
    private ObjectMapper mapper = new ObjectMapper();

    public AppKeyAndSessionProvider(
            String ssoHost, String appkey, String username, String password) {
        this.host = ssoHost;
        this.appkey = appkey;
        this.username = username;
        this.password = password;
        this.timeout = Duration.ofSeconds(30);
        // 4hrs is normal expire time
        this.sessionExpireTime = Duration.ofHours(3);
    }

    /**
     * Constructs a new session token via identity SSO. Note: These are not cached.
     *
     * @return
     * @throws IOException Thrown if authentication call fails
     * @throws InvalidCredentialException Thrown if authentication response is fail
     */
    public AppKeyAndSession getOrCreateNewSession() throws IOException, InvalidCredentialException {
        if (session != null) {
            // have a cached session - is it expired
            if ((session.getCreateTime().plus(sessionExpireTime))
                    .isAfter(Instant.now(Clock.systemUTC()))) {
                logger.info("SSO Login - session not expired - re-using");
                return session;
            } else {
                logger.info("SSO Login - session expired");
            }
        }
        logger.info("SSO Login host={}, appkey={}, username={}", host, appkey, username);
        SessionDetails sessionDetails;
        try {
            String uri =
                    String.format(
                            "https://%s/api/login?username=%s&password=%s",
                            host,
                            URLEncoder.encode(username, StandardCharsets.UTF_8),
                            URLEncoder.encode(password, StandardCharsets.UTF_8));

            Client client = Client.create();
            client.setConnectTimeout((int) (timeout.getSeconds() * 1000));
            WebResource webResource = client.resource(uri);

            ClientResponse clientResponse =
                    webResource
                            .accept("application/json")
                            .header("X-Application", appkey)
                            .header("Content-Type", "application/x-www-form-urlencoded")
                            .post(ClientResponse.class);

            mapper = new ObjectMapper();
            sessionDetails =
                    mapper.readValue(clientResponse.getEntityInputStream(), SessionDetails.class);

            logger.info("{}: Response: {}", host, sessionDetails);

        } catch (Exception e) {
            throw new IOException("SSO Authentication - call failed:", e);
        }

        // got a response - decode
        if (sessionDetails != null && "SUCCESS".equals(sessionDetails.status)) {
            session = new AppKeyAndSession(appkey, sessionDetails.token);
        } else {
            throw new InvalidCredentialException(
                    "SSO Authentication - response is fail: " + sessionDetails.error);
        }
        return session;
    }

    /** Expires cached token */
    public void expireTokenNow() {
        logger.info("SSO Login - expiring session token now");
        session = null;
    }

    public static class SessionDetails {
        public String token;
        public String product;
        public String status;
        public String error;
    }

    public Duration getTimeout() {
        return timeout;
    }

    public Duration getSessionExpireTime() {
        return sessionExpireTime;
    }
}
